### PR TITLE
Reading traits/properties from xml test-run tree using cache

### PR DESCRIPTION
**Pull request by OsirisTerje**
See #392 and PR #401

### CODE REVIEW

**Comment by CharliePoole** [comment]
I see you are using the word attributes in several places although the code had nothing to do with attributes. The adapter gets properties from NUnit and creates traits. Some properties (not all) correspond to NUnit attributes but the adapter doesn't know that.

**Comment by OsirisTerje** [comment]
Based on this line:  
      _attributesCache = new Dictionary<string, List<Trait>>();
a better name could be _traitsCache,    since the content are a keyed dictionary of traits.  
Agree?

**Comment by OsirisTerje** [comment]
All tests I have thought of have now been added and are green.  I have also tested this with comparison to the current adapter on different test code scenarios, and it returns the same.

**Comment by CharliePoole** [comment]
Yes, as well as the property.

File: src/NUnitTestAdapter/TraitsFeature.cs [hunk 1]
```diff
@@ -41,24 +40,81 @@ public static void AddTrait(this TestCase testCase, string name, string value)
...
             }
         }
 
+        private static bool IsInternalProperty(string propertyName, string propertyValue)
+        {
+            // Property names starting with '_' are for internal use only
+            return string.IsNullOrEmpty(propertyName) || propertyName[0] == '_' || string.IsNullOrEmpty(propertyValue);
+        }
+
+        private static void AddTraitsToCache(IDictionary<string, List<Trait>> propertiesCache, string key, string propertyName, string propertyValue)
```
[the remark this answers is not in the document]
Comment (OsirisTerje) [reply src/NUnitTestAdapter/TraitsFeature.cs]: Makes sense.

**Pull request merged**